### DISCUSSION TITLE
Changing tab title programmaticaly

### DISCUSSION THREAD

**Discussion by gmile** [wezterm/wezterm#2814] [General]
I am having an issue with what seems like a kindergarten question - how do I change the title of the tab, upon the tab creation?

For context, I am trying to create two workspaces, each with its own set of tabs. Tab titles are not intended to change after having initial value set. A nice-to-have would be for new tabs (spawned via Ctrl+Shift+T) must ideally follow the naming scheme, but I could live without this as long.

After searching a bit I've found this discussion https://github.com/wez/wezterm/issues/522#issuecomment-902203635 - this covers dynamic changing of title, which doesn't seem like my case (e.g. my case - simpler?).

My current config looks like this:

```lua
local wezterm = require 'wezterm'
local mux = wezterm.mux

wezterm.on('gui-startup', function()
  --[[ Set up workspace 1 ]]--
  local tab, pane, window = mux.spawn_window {
    workspace = 'api',
    args = { "nvim" },
    cwd = "/home/eugene/work/contractbok/api",
    width = 180,
    height = 40
  }

  local tab, pane, window = window:spawn_tab { cwd = '/home/eugene/work/contractbok/api' }
  tab:set_title 'tab1'

  local tab, pane, window = window:spawn_tab { cwd = '/home/eugene/work/contractbok/api' }
  tab:set_title 'tab2'

  local tab, pane, window = window:spawn_tab { cwd = '/home/eugene/work/contractbok/api' }
  tab:set_title 'tab3'

  --[[ Set up workspace 2 ]]--
  local tab, pane, window = mux.spawn_window {
    workspace = 'charts',
    args = { "nvim" },
    cwd = "/home/eugene/work/contractbok/charts",
    width = 180,
    height = 40
  }

  tab:set_title 'editor'
  local tab, pane, window = window:spawn_tab { cwd = '/home/eugene/work/contractbok/charts' }
  tab:set_title 'tab1'

  local tab, pane, window = window:spawn_tab { cwd = '/home/eugene/work/contractbok/charts' }
  tab:set_title 'tab2'

  local tab, pane, window = window:spawn_tab { cwd = '/home/eugene/work/contractbok/charts' }
  tab:set_title 'tab3'

  mux.set_active_workspace 'charts'
end)

return {}
```

It's as if I am missing something fundamental to how wezterm works, so sorry find a silly question. Pointers are more than welcome :bow:

**Comment by wez** [comment]
What exactly is the issue you're having?

Note that `tab:set_title` sets a string that can be retrieved via [tab:get_title](https://wezfurlong.org/wezterm/config/lua/MuxTab/get_title.html), but wezterm doesn't look at that string!  wezterm reads the title from the active *pane* inside of the tab in its default implementation of computing a tab title.  I know, it's a bit confusing!

You can provide your own https://wezfurlong.org/wezterm/config/lua/window-events/format-tab-title.html implementation to control exactly what goes into the tab.

**Reply by gmile** [reply]
@wez thanks, that makes sense now. I ended up being puzzled regarding when exactly the format-tab-title is being called (e.g. on every frame, on resize, etc.) and so intuitively didn't think it's the right callback to put the custom code into.

Your explanation confirmed I had a wrong impression, and it is the callback to put the custom code into. Thanks again!

**Reply by gmile** [reply]
@wez (mildly related) it's strange, but I am failing to locate a button to formally "accept" the answer using GitHub UI :thinking:

Could this be because I asked the question in **General** category and not in **Q&A**? :thinking: If yes - feel free to move this to Q&A (if there's such possibility at all exists), and I gladly accept the answer shortly afterwards. 

Never had a problem "accepting" before...

**Reply by wez** [reply]
Yeah, I think this is a discussion rather than a question. I don't know if it matters too much whether it is marked as accepted or not, so I'd suggest not worrying about it!
